src/js/common/schemaHolder.js

```diff
@@ -0,0 +1,107 @@
+    the REST API endpoints in this package.
+
+*/
+/* globals require */
+(function (fluid) {
+    "use strict";
+    if (!fluid) {
+        fluid = require("infusion");
+        fluid.require("%gpii-json-schema");
+    }
+
+    var gpii = fluid.registerNamespace("gpii");
+
+    fluid.registerNamespace("gpii.schema.schemaHolder");
+
+    gpii.schema.schemaHolder.generateIfNeeded = function (that) {
+        return that.generatedSchema ? fluid.toPromise(that.generatedSchema) : that.generateSchema();
+    };
+
+    // We cannot simply use fluid.set here because it does not return the set value, and would strip the value from the
+    // promise chain.
+    gpii.schema.schemaHolder.cacheSchema = function (that, schema) {
+        that.generatedSchema = schema;
+        return schema;
+    };
+
+    gpii.schema.schemaHolder.incorporateSubcomponentSchemas = function (that, baseSchema) {
+        var subComponents = fluid.queryIoCSelector(that, "gpii.schema.schemaHolder", true);
+        fluid.each(subComponents, function (subComponent) {
+            // getSchema is asynchronous, so we have to use a "promise chain" to handle this.
+            that.events.incorporateSubcomponentSchemas.addListener(function (baseSchema) {
+                var singleComponentSchemaPromise = fluid.promise();
+                subComponent.getSchema().then(
+                    function (subComponentSchema) {
+                        try {
+                            var mergedSchema = fluid.model.transformWithRules({ baseSchema: baseSchema, toMerge: subComponentSchema }, that.options.rules.mergeSubcomponentSchema);
+                            singleComponentSchemaPromise.resolve(mergedSchema);
+                        }
+                        catch (error) {
+                            singleComponentSchemaPromise.reject(error);
+                        }
+                    },
+                    singleComponentSchemaPromise.reject
+                );
+                return singleComponentSchemaPromise;
+            });
+        });
+        return fluid.promise.fireTransformEvent(that.events.incorporateSubcomponentSchemas, baseSchema);
+    };
+
+    fluid.defaults("gpii.schema.schemaHolder", {
+        gradeNames: ["fluid.component"],
+        mergePolicy: {
+            "rules.mergeSubcomponentSchema": "nomerge"
+        },
+        rules: {
+            // By default, simply ignore any child schemas.
+            mergeSubcomponentSchema: {
+                "": "baseSchema"
+            }
+        },
+        events: {
+            getSchema: null,
+            generateSchema: null,
+            incorporateSubcomponentSchemas: null
+        },
+        members: {
+            generatedSchema: false,
+            subComponentSchemaPromises: []
+        },
+        schema: {
+            $schema: "gss-v7-full#",
+            additionalProperties: true
+        },
+        invokers: {
+            getSchema: {
```

Comment:
whilst we can currently support events and invokers with the same name, I'd feel more comfortable if you distinguished them somehow

Reply:
That even was left over from previous work, and has been purged.

Reply:
In later discussions I noticed that there were in fact two naming collisions, and that only one had been removed.  I renamed the other.